File: module7, multiple regression/module7_3.R
toyota <- read.csv("C:/Users/USER/Downloads/ToyotaCorolla.csv")

install.packages("dplyr")
library(dplyr)

corolla <-  select(toyota,Price,Age_08_04,KM,HP,cc,Doors,Gears,Quarterly_Tax,Weight)
attach(corolla)
summary(corolla)

pairs(corolla)
cor(corolla)

#forming model taking all variables
model<-lm(Price~Age_08_04+KM+HP+cc+Doors+Gears+Quarterly_Tax+Weight, data = corolla)
summary(model)
# forming model taking individual variables
model_1<-lm(Price~Age_08_04, data = corolla)
summary(model_1)
model_2<-lm(Price~KM, data = corolla)
summary(model_2)

install.packages("GGally")
library(GGally)
ggpairs(corolla)

#for partial correlation matrix
install.packages("corpcor")
library(corpcor)
cor2pcor(cor(corolla))

#diagnostic plots
install.packages("car")
library(car)
plot(model)

#deletion diagonastic for identifying influential variable
influence.measures(model)
influenceIndexPlot(model)
influencePlot(model)

write.csv(corolla, file = "toyota_cor.csv")
getwd()

corolladata<-read.csv(file.choose())
model.1<-lm(Price~Age_08_04+KM+HP+cc+Doors+Gears+Quarterly_Tax+Weight,data=corolladata)
summary(model.1) 
summary(corolla)
gc()

#regression after deleting 81st observation
model.2<-lm(Price~Age_08_04+KM+HP+cc+Doors+Gears+Quarterly_Tax+Weight,data=corolla[-81,])
summary(model.2) #model has improved to some extent

#variance inflation factor
vif(model) #VIF > 10 = collinearity
vifage<-lm(Age_08_04~KM+HP+cc+Doors+Gears+Quarterly_Tax+Weight, data = corolla)
vifkm<-lm(KM~Age_08_04+HP+cc+Doors+Gears+Quarterly_Tax+Weight,data=corolla)
vifHP<-lm(HP~Age_08_04+KM+cc+Doors+Gears+Quarterly_Tax+Weight, data = corolla)
vifcc<-lm(cc~Age_08_04+KM+HP+Doors+Gears+Quarterly_Tax+Weight,data=corolla)
vifDoors<-lm(Doors~Age_08_04+KM+HP+cc+Gears+Quarterly_Tax+Weight,data=corolla)
vifGears<-lm(Gears~Age_08_04+KM+HP+cc+Doors+Quarterly_Tax+Weight,data=corolla)
vifqt<-lm(Quarterly_Tax~Age_08_04+KM+HP+cc+Doors+Gears+Weight,data=corolla)
vifWeight<-lm(Weight~Age_08_04+KM+HP+cc+Doors+Gears+Quarterly_Tax,data=corolla)
summary(vifage)
summary(vifkm)
summary(vifHP)
summary(vifcc)
summary(vifDoors)
summary(vifGears)
summary(vifqt)
summary(vifWeight)
#added variable plot, avplot
avPlots(model)
# from plot it is clear that Doors column has no contribution in the model so we can remove it

install.packages("MASS")
library(MASS)
stepAIC(model)
# stepAIC suggesting to built model without Doors and cc column
model.3<-lm(Price~Age_08_04+KM+HP+Gears+Quarterly_Tax+Weight, data = corolla)
summary(model.3)
model.4<-lm(Price~Age_08_04+KM+HP+Gears+Quarterly_Tax+Weight, data =corolla[-81,])
summary(model.4)
#there is no change in model after deleting 81st record so we can consider model.3 as final model
avPlots(model.3)

#trainig and testing model
n=nrow(corolla)
n1=n*0.7
n1
train=sample(1:n,n1)
train
test=corolla[-train,]
test
pred=predict(model.3,newdata = test)
pred
actual=test$Price
actual
error = actual-pred
error
test.rmse=sqrt(mean(error**2))
test.rmse
train.rmse=sqrt(mean(model.3$residuals^2))
train.rmse

File: module7, multiple regression/module7_3.py
# -*- coding: utf-8 -*-
"""
Created on Sun Apr 26 00:49:19 2020

@author: USER
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

toyota=pd.read_csv("C:/Users/USER/Downloads/ToyotaCorolla.csv", encoding='latin')
corolla=toyota[["Price","Age_08_04","KM","HP","cc","Doors","Gears","Quarterly_Tax","Weight"]]

corolla.corr()

import pandas as pd
# for displaying entire correlation matrix in console
#pd.set_option('display.height', 1000)
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

import seaborn as sns
sns.pairplot(corolla)

#preparing model considering all variables
import statsmodels.formula.api as smf # for regression model
model=smf.ols("Price~Age_08_04+KM+HP+cc+Doors+Gears+Quarterly_Tax+Weight", data=corolla).fit()
model.summary()
model_1=smf.ols("Price~Age_08_04+KM+HP+cc+Gears+Quarterly_Tax+Weight", data=corolla).fit()
model_1.summary()
model_2=smf.ols("Price~Doors",data=corolla).fit()
model_2.summary()

#influence plot
import statsmodels.api as sm
sm.graphics.influence_plot(model)

t_corolla=corolla.drop(corolla.index[[-80]],axis=0) #dropping 80th record
model_3=smf.ols("Price~Age_08_04+KM+HP+cc+Doors+Gears+Quarterly_Tax+Weight",data=t_corolla).fit()
model_3.summary() #no improvement in model instead it is getting slightly worse

#calculating vif values of independent variables
rsq_Age_08_04=smf.ols("Age_08_04~KM+HP+cc+Doors+Gears+Quarterly_Tax+Weight", data=corolla).fit().rsquared
vif_Age_08_04=1/(1-rsq_Age_08_04)
rsq_KM=smf.ols("KM~Age_08_04+HP+cc+Doors+Gears+Quarterly_Tax+Weight", data=corolla).fit().rsquared
vif_KM=1/(1-rsq_KM)
rsq_HP=smf.ols("HP~Age_08_04+KM+cc+Doors+Gears+Quarterly_Tax+Weight", data=corolla).fit().rsquared
vif_HP=1/(1-rsq_HP)
rsq_cc=smf.ols("cc~Age_08_04+HP+KM+Doors+Gears+Quarterly_Tax+Weight", data=corolla).fit().rsquared
vif_cc=1/(1-rsq_cc)
rsq_Doors=smf.ols("Doors~Age_08_04+HP+cc+KM+Gears+Quarterly_Tax+Weight", data=corolla).fit().rsquared
vif_Doors=1/(1-rsq_Doors)
rsq_Gears=smf.ols("Gears~Age_08_04+HP+cc+Doors+KM+Quarterly_Tax+Weight", data=corolla).fit().rsquared
vif_Gears=1/(1-rsq_Gears)
rsq_Quarterly_Tax=smf.ols("Quarterly_Tax~Age_08_04+HP+cc+Doors+Gears+KM+Weight", data=corolla).fit().rsquared
vif_Quarterly_Tax=1/(1-rsq_Quarterly_Tax)
rsq_Weight=smf.ols("Weight~Age_08_04+HP+cc+Doors+Gears+Quarterly_Tax+KM", data=corolla).fit().rsquared
vif_Weight=1/(1-rsq_Weight)

d1={"variables":["Age_08_04","KM","HP","cc","Doors","Gears","Quarterly_Tax","Weight"],"vif":[vif_Age_08_04,vif_KM,vif_HP,vif_cc,vif_Doors,vif_Gears,vif_Quarterly_Tax,vif_Weight]}
VIF_frame=pd.DataFrame(d1)
VIF_frame # all vif values are below 10 means no multicollinearity problem

corolla.describe()#for summary of dataset
_ = plt.hist(corolla['KM'])

corolla.corr()

model_4=smf.ols("Price~Age_08_04+KM+HP+Weight", data=corolla).fit()
model_4.summary()
sm.graphics.plot_partregress_grid(model_4)

#added variable plot
sm.graphics.plot_partregress_grid(model)
#as we can see from this plot door column is not contributing anything into the model so will proceed removing this column

#making model without door column and cc column
#final model
model_1=smf.ols("Price~Age_08_04+KM+HP+Gears+Quarterly_Tax+Weight", data=corolla).fit()
model_1.summary()
sm.graphics.plot_partregress_grid(model_1)

#predicted value for price
price_pred=model_1.predict(corolla[["Age_08_04","KM","HP","Gears","Quarterly_Tax","Weight"]])
price_pred

#observed value vs fitted value
plt.scatter(corolla.Price,price_pred,c='r');plt.xlabel("observed value");plt.ylabel("fitted value")

#residuals vs fitted value
plt.scatter(price_pred,model_1.resid_pearson, c="r"),plt.axhline(y=0,color='blue');plt.xlabel('fitted value');plt.ylabel('residuals')

#normality plot for residuals
#histogram
plt.hist(model_1.resid_pearson)

#qq plot for residuals
import pylab
import scipy.stats as st

#checking residuals are normally distributed
st.probplot(model_1.resid_pearson, dist="norm",plot=pylab)

### Splitting the data into train and test data 
from sklearn.model_selection import train_test_split
corolla_train,corolla_test  = train_test_split(corolla,test_size = 0.3) # 30% test data
corolla_train
corolla_test

# preparing the model on train data 
model_train=smf.ols("Price~Age_08_04+KM+HP+Gears+Quarterly_Tax+Weight", data=corolla_train).fit()

# train data prediction
train_pred = model_train.predict(corolla_train)

# train residual values 
train_resid  = train_pred - corolla_train.Price

# RMSE value for train data 
train_rmse = np.sqrt(np.mean(train_resid*train_resid)) #1273.433489490905
train_rmse

# prediction on test data set 
test_pred = model_train.predict(corolla_test)

# test residual values 
test_resid  = test_pred - corolla_test.Price

# RMSE value for test data 
test_rmse = np.sqrt(np.mean(test_resid*test_resid)) #1488.1445671977026
test_rmse


















































 



























